### PR TITLE
API endpoints for ping, models, workflows & runs (with authentication)

### PR DESCRIPTION
**Pull request by tezansahu**
## Description

Created API endpoints for ping, models, workflows & runs (with Basic HTTP authentication), along with the OpenAPI specifications for the Swagger UI.

## Motivation and Context

As the part of my GSoC Project: Extending the PEcAn API, this is my first PR incorporating some of the basic GET endpoints for the PEcAn Project.

## Review Time Estimate

- [ ] Immediately
- [ ] Within one week
- [x] When possible
## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue) 
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] My change requires a change to the documentation.
- [x] I have updated the CHANGELOG.md.
- [ ] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

File: apps/api/auth.R [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+#* @param crypt_pass Encrypted password
+#* @return TRUE if encrypted password is correct, else FALSE
+#* @author Tezan Sahu
+validate_crypt_pass <- function(username, crypt_pass) {
+  settings <-list(database = list(bety = list(
+    driver = "PostgreSQL", 
+    user = "bety", 
+    dbname = "bety", 
+    password = "bety", 
+    host="postgres"
```
Comment (istfer) [inline apps/api/auth.R:34]: are you using docker? couldn't host be different, maybe it is best not to have hardcoded db settings?

Comment (tezansahu) [reply apps/api/auth.R:34]: As of now, to run locally (without docker), you can just modify `host="postgres` to `host=localhost`.

Comment (tezansahu) [reply apps/api/auth.R:34]: @robkooper Do let me know your thoughts about how we could parameterize the host if possible.

Comment (tezansahu) [reply apps/api/auth.R]: Changed db connection at all points to use `betyConnect()`

File: apps/api/auth.R [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+  )))
+  dbcon <- PEcAn.DB::db.open(settings$database$bety)
```
[the remark this answers is not in the document]
Comment (tezansahu) [reply apps/api/auth.R]: Changed db connection at all points to use `betyConnect()`

File: apps/api/entrypoint.R [hunk 3]
```diff
@@ -0,0 +1,25 @@
+source("auth.R")
+source("ping.R")
+
+root <- plumber::plumber$new()
+root$setSerializer(plumber::serializer_unboxed_json())
+
+root$filter("require-auth", authenticate_user)
+
+root$handle("GET", "/api/ping", ping)
```
Comment (robkooper) [inline apps/api/entrypoint.R:9]: can you document here why you use this syntax vs the plumber$new

Comment (tezansahu) [reply apps/api/entrypoint.R:17]: Added comments to the `entrypoint.R` & other relevant files

**Comment by robkooper** [comment]
@mdietze, Tezan and I discussed this both in our weekly meeting, as well as in slack. We decided to take the code to split in 2 pieces, one is the rest endpoint in /apps/api, the second part is the actual user library.

This pull request, is the endpoint that people can talk to use PEcAn remotely. This requires no access to the database from the users perspective (api code does access the database), and they can use this API to work with PEcAn.

The second piece that will replace most of the code @ashiklom wrote (that lives in /api) will go in its own repository. This is not done yet. The goal of this new library, is for somebody to install and work with PEcAn without access to the database.

As for the reason for the /apps/api folder is that this is not a library, but an application that we will deploy (plumbr does not make it easy to have it as a library). The thinking is that if we have other PEcAn applications, we can put them in this folder, akin to the shiny folder.

**Comment by mdietze** [comment]
@robkooper that plan makes sense (though I'd encourage trying to build on what @ashiklom did rather than starting from scratch). Make sure to deprecate the old API folder so that other users/developers are also not confused by the presence of two separate API folders.

@tezansahu this PR still needs documentation and tests.

**Comment by tezansahu** [comment]
Sure @mdietze I'm working on the documentation and will push it my tomorrow. I have also been working on dockerizing the API application so that it can be served as a container. All this will be reflected in my next push.

**Comment by tezansahu** [comment]
In the current documentation, the link for OpenAPI 3 Documentation of the PEcAn API is a temporary link (which may not be directly usable to execute API calls at this point). This will be updated as soon as we host the documentation and the API server somewhere (yet to be decided).

File: book_source/03_topical_pages/07_remote_access/01_pecan_api.Rmd [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+This page contains the high-level overviews & the functionalities offered by the different RESTful endpoints of the PEcAn API. For more detailed information including the request & response structure along with relevant schemas, you can visit the [OpenAPI 3 Documentation for the PEcAn API](https://petstore.swagger.io/?url=https://raw.githubusercontent.com/PecanProject/pecan/299c240277d616463979cdbe09ef57731b8167fc/apps/api/pecanapi-spec.yml).
+
+The currently implemented functionalities include:
+
+* __General:__
+  * `GET /api/ping`: Ping the server to check if it is live
+  * `GET /api/status`: Obtain general information about PEcAn & the details of the database host
+
+* __Models:__
+  * `GET /api/models/`: Retrieve the details of environmental model(s) used by PEcAn (can be filtered based on `model_name` & `revision`)
```
Comment (mdietze) [inline book_source/03_topical_pages/07_remote_access/01_pecan_api.Rmd:33]: Drop "environmental". Explain how to filter based on model_name and revision.

File: book_source/03_topical_pages/07_remote_access/01_pecan_api.Rmd [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+
+* __Workflows:__
+  * `GET /api/workflows/`: Retrieve a list of PEcAn workflows (can be filtered based on `model_id` & `site_id`)
```
Comment (mdietze) [inline book_source/03_topical_pages/07_remote_access/01_pecan_api.Rmd:36]: likewise, how do you filter?

File: book_source/03_topical_pages/07_remote_access/01_pecan_api.Rmd [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+  * `POST /api/workflows/` *: Submit a new PEcAn workflow
+  * `GET /api/workflows/{id}`: Obtain the details of a particular PEcAn workflow by supplying its ID
+
+* __Runs:__
+  * `GET /api/runs`:  Get the list of all the runs for a specified PEcAn Workflow
```
Comment (mdietze) [inline book_source/03_topical_pages/07_remote_access/01_pecan_api.Rmd:41]: How do you specify the workflow ID? The obvious thing to try would be to add the workflow ID, but the next line makes it clear that that syntax would look up the run ID not the workflow ID

Comment (tezansahu) [reply book_source/03_topical_pages/07_remote_access/01_pecan_api.Rmd:41]: @mdietze all these specifications are mentioned in the openapi3 documentation mentioned in the link. That is why I have kept this documentation very minimalistic

Comment (tezansahu) [reply book_source/03_topical_pages/07_remote_access/01_pecan_api.Rmd:41]: This includes all schemas, parameter usage, etc. As of now, u can try to visit the temporary link mentioned in the documentation to see the detailed API documentation

Comment (mdietze) [reply book_source/03_topical_pages/07_remote_access/01_pecan_api.Rmd:41]: If the primary documentation is elsewhere then that needs to be much more clear -- this is going to be missed by too many users. Likewise, ecologists are going to need some clear worked examples of code that uses these APIs (e.g. R and Python), even if that documentation gets updated later with additional examples using your planned package.

File: book_source/03_topical_pages/07_remote_access/01_pecan_api.Rmd (hunk 4 above)
Comment (tezansahu) [reply book_source/03_topical_pages/07_remote_access/01_pecan_api.Rmd]: Dropped "environmental". I have also added examples in towards the later part of this documentation that will show users how to filter based on `model_name` & `revision`

File: book_source/03_topical_pages/07_remote_access/01_pecan_api.Rmd (hunk 5 above)
Comment (tezansahu) [reply book_source/03_topical_pages/07_remote_access/01_pecan_api.Rmd]: I have also added examples in towards the later part of this documentation that will show users how to filter based on `model_id` & `site_id`

File: book_source/03_topical_pages/07_remote_access/01_pecan_api.Rmd (hunk 6 above)
Comment (tezansahu) [reply book_source/03_topical_pages/07_remote_access/01_pecan_api.Rmd:43]: I have made the line indicating the link to detailed Swagger Documentation bold & clear, and have also added some simple examples towards the end of this page that tell the user how to interact with the PEcAn API server once hosted

**Pull request merged**